lib/package.gi

```diff
@@ -159,6 +159,159 @@ InstallGlobalFunction( SetPackageInfo, function( record )
+    if file = fail then
+      files := [];
+      # Perhaps some subdirectories contain `PackageInfo.g' files.
+      for subdir in Set( DirectoryContents( pkgpath ) ) do
+        if not subdir in [ ".", ".." ] then
+          pkgpath:= Filename( [ pkgdir ],
+                              Concatenation( name, "/", subdir ) );
+          if pkgpath <> fail and IsDirectoryPath( pkgpath )
+                              and not subdir in [ ".", ".." ] then
+            file:= Filename( [ pkgdir ],
+                Concatenation( name, "/", subdir, "/PackageInfo.g" ) );
+            if file <> fail then
+              Add( files,
+                    [ file, Concatenation( name, "/", subdir ) ] );
+            fi;
+          fi;
+        fi;
+      od;
+    else
+      files:= [ [ file, name ] ];
+    fi;
+    return files;
+end );
+
+#############################################################################
+##
+#F  EnrichMetapackageRecord( subpackagerecord, metapackagerecord )
+##
+
+BindGlobal( "EnrichMetapackageRecord", function( subpackagerecord, metapackagerecord )
+    local version, name;
+    if not IsBound( metapackagerecord.Dependencies.NeededOtherPackages ) then
+        metapackagerecord.Dependencies.NeededOtherPackages := [];
+    fi;
+
+    name := subpackagerecord.PackageName;
+    version := subpackagerecord.Version;
+    version := Concatenation( "= ", version );
+
+    Add( metapackagerecord.Dependencies.NeededOtherPackages, [ name, version ] );
+
+end );
+
+#############################################################################
+##
+#F  EnrichSubpackageRecord( subpackagerecord, metapackagerecord )
+##
+
+BindGlobal( "EnrichSubpackageRecord", function( subpackagerecord, metapackagerecord )
+    if not IsBound( subpackagerecord.SubpackageOf ) then
+        subpackagerecord.SubpackageOf := metapackagerecord.PackageName;
+    fi;
+end );
+
+#############################################################################
+##
+#F  AddPackageInfo( files )
+##
+InstallGlobalFunction( AddPackageInfos, function( files, pkgdir, ignore, metapackagerecord... )
+    local file, record, pkgname, version, is_metapackage, name;
+    if Length( metapackagerecord ) = 1 then
+      is_metapackage := true;
+      metapackagerecord := metapackagerecord[ 1 ];
+    else
+      is_metapackage := false;
+    fi;
+
+    for file in files do
+      # Read the `PackageInfo.g' file.
+      Unbind( GAPInfo.PackageInfoCurrent );
+      Read( file[1] );
+      if IsBound( GAPInfo.PackageInfoCurrent ) then
+        record:= GAPInfo.PackageInfoCurrent;
+        Unbind( GAPInfo.PackageInfoCurrent );
+        pkgname:= LowercaseString( record.PackageName );
+        NormalizeWhitespace( pkgname );
+        version:= record.Version;
+
+        # If we have this version already then leave it out.
+        if ForAll( GAPInfo.PackagesInfo,
+                    r ->    r.PackageName <> record.PackageName
+                        or r.Version <> version ) then
```

Comment:
Optional: I know this is old code, but: once again, I'd prefer if it looked like this (and IMHO, that'd also match the comment before it better):
```gap
if ForAny( GAPInfo.PackagesInfo, r -> [r.PackageName, r.Verson] = [record.PackageName, verson] ) then
  continue;
fi;